src/Generator.cpp

```diff
@@ -21,362 +26,655 @@
+                }
+            }
+            else {
+                if (parameters_.generation().gun_type == "E")
+                    energy_incident = parameters_.generation().energy;
+                else
+                    energy_incident = parameters_.generation().energy*((exp(eta_incident*eta_incident)+1)/(exp(eta_incident*eta_incident)-1));
+            }
+
+            int position = &ip-&parameters_.general().part_type[0];
+            event.fillGenEn(position, energy_incident);
+            event.fillGenEta(position, eta_incident);
+            event.fillGenPhi(position, phi_incident);
+            event.fillPDGid(position, ip);
+
+            // Number of hits
+            // as the hits are not generated yet, the energy is computed with the smaller energy
+            // resolution in order to have the bigger number of hits per events
+            // energy spot
+            // no fluctuations: fixed energy = 1. / nhitspergev
+            // fluctuations: alpha/sqrt(E) -> Poissonian nbr hits of energy 1/alpha^2
+            // where alpha is the stochastic term of the resolution
+
+            for (int layer_id = layer_min; layer_id < layer_max; layer_id++) {
+
+                // incident direction
+                double thetainc = 2.*std::atan(std::exp(-eta_incident));
+                double r = parameters_.geometry().layers_z[layer_id]*tan(thetainc); 
+                double incident_x = r*cos(phi_incident);
+                double incident_y = r*sin(phi_incident);
+
+                double r0_electro = aShowerParametrization.r0_electro_(layer_id, ip);
+                double r0_hadro = aShowerParametrization.r0_hadro_(layer_id, ip);
+
+                // take longitudinal profile as mean energy per layer for fixed energy
+                double layer_weight = aShowerParametrization.getLayerProfile(ip)[layer_id];
+
+                double denrj;
+                int nhits;
+                if (!parameters_.generation().fluctuation){
+                    denrj = 1./parameters_.generation().number_of_hits_per_gev;
+                    nhits = int(energy_incident*layer_weight*parameters_.generation().number_of_hits_per_gev);
+                }
+                else {
+                    denrj = aShowerParametrization.spotEnergy(ip)[2];
+                    nhits = gun_.Poisson(energy_incident*layer_weight/denrj);
+                }
+                tot_hits += nhits;
+
+                double real_energy = 0;
+
+                // loop over hits
+                for (int i = 0; i<nhits; i++) {
+                    // Compute the position of each hit
+
+                    double r_shower = 0;
+                    if (ip == 11 || ip == 22)
+                        r_shower = gun_.Exp(r0_electro); // exponential exp(-r/r0)
+                    else // for hadronic shower templates parametrization
+                        r_shower = 108 * gun_.Exp(r0_hadro/19.3) + 95 * gun_.Exp(r0_hadro/76.);
```

Comment:
Remove magic numbers 108, 19.3, 95, 76, and use constant variables or input parameters instead

Reply:
Fixed